network/containerizer/shim.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package containerizer
+
+import (
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/state"
+)
+
+// LinkLayerDevice is an indirection for state.LinkLayerDevice.
+// It facilitates testing the provisioner's use of this package.
+type LinkLayerDevice interface {
+	Name() string
+	Type() state.LinkLayerDeviceType
+	MACAddress() string
+	ParentName() string
+	ParentDevice() (LinkLayerDevice, error)
+	EthernetDeviceForBridge(name string) (state.LinkLayerDeviceArgs, error)
+	Addresses() ([]*state.Address, error)
+
+	// These are recruited in tests. See comment on Machine below.
+	MTU() uint
+	IsUp() bool
+	IsAutoStart() bool
+}
+
+// linkLayerDevice implements LinkLayerDevice.
+// We need our own implementation of the indirection above in order to mock the
+// return of ParentDevice, which in the state package returns a reference to a
+// raw state.LinkLayerDevice.
+type linkLayerDevice struct {
+	*state.LinkLayerDevice
+}
+
+// ParentDevice implements LinkLayerDevice by wrapping the response from the
+// inner device's call in a new instance of linkLayerDevice.
+func (l *linkLayerDevice) ParentDevice() (LinkLayerDevice, error) {
+	dev, err := l.LinkLayerDevice.ParentDevice()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &linkLayerDevice{dev}, nil
+}
+
+var _ LinkLayerDevice = (*linkLayerDevice)(nil)
+
+// Machine is an indirection for state.Machine,
+// describing a machine that is to host containers.
+type Machine interface {
+	Id() string
+	AllSpaces() (set.Strings, error)
+	LinkLayerDevicesForSpaces([]string) (map[string][]LinkLayerDevice, error)
+	SetLinkLayerDevices(devicesArgs ...state.LinkLayerDeviceArgs) (err error)
+
+	// TODO (manadart 2018-10-10) These methods are used in tests, which rely
+	// on the StateSuite. Some of them are recruited via the Container
+	// interface below, but they are all located here for simplicity.
+	// A better approach could be sought that does not require their
+	// presence here.
+	SetDevicesAddresses(devicesAddresses ...state.LinkLayerDeviceAddress) (err error)
+	SetParentLinkLayerDevicesBeforeTheirChildren(devicesArgs []state.LinkLayerDeviceArgs) error
+	AllLinkLayerDevices() ([]LinkLayerDevice, error)
+	SetConstraints(cons constraints.Value) (err error)
+	RemoveAllAddresses() error
+	Wrapped() *state.Machine
+}
+
+// MachineShim implements Machine.
+// It is required to mock the return of LinkLayerDevicesForSpaces,
+// which includes raw state.LinkLayerDevice references.
```

Comment:
Objects returning Objects is the one case where Go's interface mechanism tends to fail. They chose not to do recursive interfaces, but so often it would be nice if they did.